kv/kvserver: TestReadLoadMetricAccounting failed

kv/kvserver.TestReadLoadMetricAccounting [failed](https://tanzanite.cluster.engflow.com/invocations/default/1e1b94cd-e0db-422b-adc2-bf96dca9f19d?testReportRun=3&testReportShard=27&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [416e2691fd0f0cbde1931905e8aefbb7bbfb39a7](https://github.com/cockroachdb/cockroach/commits/416e2691fd0f0cbde1931905e8aefbb7bbfb39a7):


```
=== RUN   TestReadLoadMetricAccounting
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReadLoadMetricAccounting2713074910
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_rankings_test.go:206: 
        	Error Trace:	pkg/kv/kvserver/replica_rankings_test.go:206
        	            				pkg/kv/kvserver/replica_rankings_test.go:500
        	Error:      	Max difference between 0 and 85 allowed is 4, but difference was -85
        	Test:       	TestReadLoadMetricAccounting
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestReadLoadMetricAccounting2713074910
--- FAIL: TestReadLoadMetricAccounting (4.74s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=27</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141586 kv/kvserver: TestReadLoadMetricAccounting failed [A-testing C-bug C-test-failure O-robot P-3 T-kv branch-release-25.1]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadLoadMetricAccounting.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

kv/kvserver.TestReadLoadMetricAccounting [failed](https://tanzanite.cluster.engflow.com/invocations/default/38ec11c2-e547-4c95-b6a9-280b2ef836b4?testReportRun=2&testReportShard=27&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [dece7feb5f1c957bbc7714afca87550895316d44](https://github.com/cockroachdb/cockroach/commits/dece7feb5f1c957bbc7714afca87550895316d44):


```
=== RUN   TestReadLoadMetricAccounting
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReadLoadMetricAccounting1776015150
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_rankings_test.go:411: evaluated [logtags=: *kvpb.AddSSTableRequest
    replica_rankings_test.go:414: evaluated [logtags=]: *kvpb.RequestLeaseRequest on /Table/Max: header:<key:"\372" > lease:<start:<> replica:<node_id:1 store_id:1 replica_id:1 type:VOTER_FULL > proposed_ts:<wall_time:1740145404035220521 > sequence:1 acquisition_type:Request min_expiration:<wall_time:1740145410035220521 > term:6 > prev_lease:<start:<> expiration:<wall_time:1740145409956575575 > replica:<node_id:1 store_id:1 replica_id:1 type:VOTER_FULL > deprecated_start_stasis:<wall_time:1740145409956575575 > proposed_ts:<wall_time:1740145403956575575 > sequence:1 acquisition_type:Request min_expiration:<> >  {Timestamp:1740145404.036308891,0 TimestampFromServerClock:<nil> Now:0,0 Replica:(n0,s0):? RangeID:75 UserPriority:normal Txn:<nil> ReadConsistency:CONSISTENT RoutingPolicy:LEASEHOLDER WaitPolicy:Block LockTimeout:0s MaxSpanRequestKeys:0 TargetBytes:0 WholeRowsOfSize:0 AllowEmpty:false ReturnOnRangeBoundary:false DistinctSpans:false AsyncConsensus:false CanForwardReadTimestamp:false GatewayNodeID:0 ClientRangeInfo:{DescriptorGeneration:0 LeaseSequence:0 ClosedTimestampPolicy:LAG_BY_CLUSTER_SETTING ExplicitlyRequested:false} BoundedStaleness:<nil> TraceInfo:<nil> IndexFetchSpec:<nil> ReturnElasticCPUResumeSpans:false ProfileLabels:[] AmbiguousReplayProtection:false ConnectionClass:default ProxyRangeInfo:<nil> WriteOptions:<nil> DeadlockTimeout:0s}
    replica_rankings_test.go:502: test #1
    replica_rankings_test.go:414: evaluated [logtags=]: *kvpb.GetRequest on �000000000: header:<key:"\372000000000" >  {Timestamp:1740145404.042004351,0 TimestampFromServerClock:1740145404.042004351,0 Now:1740145404.041849601,0 Replica:(n1,s1):1 RangeID:75 UserPriority:normal Txn:<nil> ReadConsistency:CONSISTENT RoutingPolicy:LEASEHOLDER WaitPolicy:Block LockTimeout:0s MaxSpanRequestKeys:0 TargetBytes:0 WholeRowsOfSize:0 AllowEmpty:false ReturnOnRangeBoundary:false DistinctSpans:false AsyncConsensus:false CanForwardReadTimestamp:false GatewayNodeID:1 ClientRangeInfo:{DescriptorGeneration:1 LeaseSequence:1 ClosedTimestampPolicy:LAG_BY_CLUSTER_SETTING ExplicitlyRequested:false} BoundedStaleness:<nil> TraceInfo:<nil> IndexFetchSpec:<nil> ReturnElasticCPUResumeSpans:false ProfileLabels:[] AmbiguousReplayProtection:false ConnectionClass:default ProxyRangeInfo:<nil> WriteOptions:<nil> DeadlockTimeout:0s}
    replica_rankings_test.go:209: 
        	Error Trace:	pkg/kv/kvserver/replica_rankings_test.go:209
        	            				pkg/kv/kvserver/replica_rankings_test.go:530
        	Error:      	Max difference between 0 and 85 allowed is 4, but difference was -85
        	Test:       	TestReadLoadMetricAccounting
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestReadLoadMetricAccounting1776015150
--- FAIL: TestReadLoadMetricAccounting (5.26s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=27</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadLoadMetricAccounting.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -2309,9 +2309,7 @@ func TestRebalancingAndCrossRegionZoneSnapshotMetrics(t *testing.T) {
 	// an epoch lease. For leader leases, however, the range starts off with no
 	// leader so the first lease that's proposed is an expiration lease, which
 	// gets upgraded to a leader lease once a leader is elected.
-	if !kvserver.ExpirationLeasesOnly.Get(&tc.Server(0).ClusterSettings().SV) {
-		tc.WaitForLeaseUpgrade(ctx, t, desc)
-	}
+	tc.MaybeWaitForLeaseUpgrade(ctx, t, desc)
 	// sendSnapshotFromServer is a testing helper that sends a learner snapshot
 	// from server[0] to server[serverIndex] and returns the expected size (in
 	// bytes) of the snapshot sent.

--- a/pkg/kv/kvserver/replica_rankings_test.go
+++ b/pkg/kv/kvserver/replica_rankings_test.go
@@ -497,6 +497,10 @@ func TestReadLoadMetricAccounting(t *testing.T) {
 	// Disable the consistency checker, to avoid interleaving requests
 	// artificially inflating measurement due to consistency checking.
 	sqlDB.Exec(t, `SET CLUSTER SETTING server.consistency_check.interval = '0'`)
+	// Wait for lease upgrade, to avoid interleaving upgrade requests inflating
+	// the measurements below.
+	desc := tc.LookupRangeOrFatal(t, scratchKey)
+	tc.MaybeWaitForLeaseUpgrade(ctx, t, desc)
 
 	for i, testCase := range testCases {
 		t.Logf("test #%d", i+1)

--- a/pkg/testutils/serverutils/test_cluster_shim.go
+++ b/pkg/testutils/serverutils/test_cluster_shim.go
@@ -185,6 +185,11 @@ type TestClusterInterface interface {
 		t TestFataler, rangeDesc roachpb.RangeDescriptor, dest roachpb.ReplicationTarget,
 	)
 
+	// MaybeWaitForLeaseUpgrade waits until the lease held for the given range
+	// descriptor is upgraded to an epoch-based or leader lease, but only if we
+	// expect the lease to be upgraded.
+	MaybeWaitForLeaseUpgrade(_ context.Context, _ TestFataler, _ roachpb.RangeDescriptor)
+
 	// MoveRangeLeaseNonCooperatively performs a non-cooperative transfer of the
 	// lease for a range from whoever has it to a particular store. That store
 	// must already have a replica of the range. If that replica already has the

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -1157,8 +1157,8 @@ func (tc *TestCluster) TransferRangeLeaseOrFatal(
 }
 
 // MaybeWaitForLeaseUpgrade waits until the lease held for the given range
-// descriptor is upgraded to an epoch-based one, but only if we expect the lease
-// to be upgraded.
+// descriptor is upgraded to an epoch-based or leader lease, but only if we
+// expect the lease to be upgraded.
 func (tc *TestCluster) MaybeWaitForLeaseUpgrade(
 	ctx context.Context, t serverutils.TestFataler, desc roachpb.RangeDescriptor,
 ) {

